c-scape/src/rand/mod.rs

```diff
@@ -0,0 +1,57 @@
+use libc::{c_int, c_uint};
+
+use rand::Rng;
+use rand_core::SeedableRng;
+use rand_pcg::Pcg32;
+
+use core::cell::SyncUnsafeCell;
+
+#[cfg(test)]
+static_assertions::assert_eq_size!(c_uint, u32);
+
+// We use PCG32 here, which takes just 16 bytes and offers us a relatively nice
+// non-cryptographic RNG for most applications.
+// TODO: Remove the option once Pcg32 has seedable const fn things
+static STATE: SyncUnsafeCell<Option<Pcg32>> = SyncUnsafeCell::new(None);
+
+#[no_mangle]
+unsafe extern "C" fn rand() -> c_int {
+    libc!(libc::rand());
+
+    // POSIX requires that if the user hasn't initialized the RNG, it behaves
+    // as-if srand(1) was called.
+    if *STATE.get() == None {
+        srand(1);
+    }
+
+    // Sample using a uniform distribution
+    (*STATE.get())
```

Comment:
`rand` is supposed to be MT-Safe. Accessing `STATE` from multiple threads at the same time without synchronization is UB.

Reply:
I think being safe to call in the presence of threads means that two threads can call `rand` at the same time without UB. I think it makes sense to add the lock here.